src/main/java/hudson/tasks/test/AbstractTestResultAction.java

```diff
@@ -287,29 +336,235 @@ public TestResult findCorrespondingResult(String id) {
+                projectCount++;
+            }
+            if (projectCount > projectCountLimit) {
+                int levelToDel = projectSet.size() - 1;
+                int deletionSize = projectSet.get(levelToDel).size();
+                projectSet.remove(levelToDel);
+                projectCount -= deletionSize;
+                levelCap = projectSet.size() - 1;
+            }
+        }
+
+        /*
+         * Converting the set of package names to the corresponding array
+         * and finally sorting it in ascending order.
+         */
+        projectList = new String[projectCount];
+        int index = 0;
+        for (int level : projectSet.keySet()) {
+            Set<String> projectSetPerLevel = projectSet.get(level);
+            for (String projectName : projectSetPerLevel) {
+                projectList[index] = projectName;
+                index++;
+            }
+        }
+        Arrays.sort(projectList);
+        return projectList;
+    }
+
     /**
      * Generates a PNG image for the test result trend.
      */
-    public void doGraph( StaplerRequest req, StaplerResponse rsp) throws IOException {
-        if(ChartUtil.awtProblemCause!=null) {
+    public void doGraph (StaplerRequest req, StaplerResponse rsp) throws IOException {
+        if (ChartUtil.awtProblemCause != null) {
             // not available. send out error message
-            rsp.sendRedirect2(req.getContextPath()+"/images/headless.png");
+            rsp.sendRedirect2(req.getContextPath() + "/images/headless.png");
             return;
         }
 
-        if(req.checkIfModified(run.getTimestamp(),rsp))
+        if (req.checkIfModified(run.getTimestamp(), rsp))
             return;
 
-        ChartUtil.generateGraph(req,rsp,createChart(req,buildDataSet(req)),calcDefaultSize());
+        /*
+         * Utility method for creating various test result trends.
+         */
+        doGraphUtil(req, rsp);
+    }
+
+    /**
+     * A utility method for constructing trends based upon the query parameters passed in the
+     * request message.
+     * @param req HTTP request message for the image of trend.
+     * @param rsp HTTP response message for the requested image.
+     * @throws IOException in case an exception occurs in
+     * {@link ChartUtil#generateGraph(StaplerRequest, StaplerResponse, JFreeChart, Area)}
+     */
+    public void doGraphUtil (StaplerRequest req, StaplerResponse rsp) throws IOException {
+        String projectLevel = getParameter(req, PROJECTLEVEL);
+        String trendType = getParameter(req, TRENDTYPE);
+
+        /*
+         * A binary search for verifying whether the given project level is valid or not. If found in the
+         * array or is equal to "AllProjects" the valid else not.
+         */
+        int index = Arrays.binarySearch(projectList, projectLevel);
+        if ((index >= 0 || projectLevel.equals(ALLPROJECTS)) && trendType.equals(BUILD_ANALYSIS)) {
```

Comment:
Similarly here `(index >= 0 || projectLevel.equals(ALLPROJECTS))` can be clubbed together and use switch.
Same comment for indentation

Reply:
Done.